pkg/controller/linstorcsidriver/linstorcsidriver_controller.go

```diff
@@ -0,0 +1,878 @@
+			{APIGroups: []string{""}, Resources: []string{"secrets"}, Verbs: []string{"get", "list"}},
+			{APIGroups: []string{"snapshot.storage.k8s.io"}, Resources: []string{"volumesnapshotclasses"}, Verbs: []string{"get", "list", "watch"}},
+			{APIGroups: []string{"snapshot.storage.k8s.io"}, Resources: []string{"volumesnapshotcontents"}, Verbs: []string{"create", "get", "list", "watch", "update", "delete"}},
+			{APIGroups: []string{"snapshot.storage.k8s.io"}, Resources: []string{"volumesnapshots"}, Verbs: []string{"get", "list", "watch", "update"}},
+			{APIGroups: []string{"snapshot.storage.k8s.io"}, Resources: []string{"volumesnapshots/status"}, Verbs: []string{"update"}},
+			{APIGroups: []string{"apiextensions.k8s.io"}, Resources: []string{"customresourcedefinitions"}, Verbs: []string{"create", "list", "watch", "delete", "get", "update"}},
+		},
+	}
+}
+
+func newCSIPriorityClass(cr *piraeusv1alpha1.LinstorCSIDriver) *schedv1.PriorityClass {
+	return &schedv1.PriorityClass{
+		ObjectMeta:    makeMeta(cr, PriorityClass),
+		Value:         1000000,
+		GlobalDefault: false,
+		Description:   "Priority class for piraeus-csi components",
+	}
+}
+
+func newCSIControllerServiceAccount(cr *piraeusv1alpha1.LinstorCSIDriver) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: makeMeta(cr, ControllerServiceAccount),
+	}
+}
+
+func newCSINodeServiceAccount(cr *piraeusv1alpha1.LinstorCSIDriver) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: makeMeta(cr, NodeServiceAccount),
+	}
+}
+
+func newCSINodeDaemonSet(cr *piraeusv1alpha1.LinstorCSIDriver) *appsv1.DaemonSet {
+	return &appsv1.DaemonSet{
+		ObjectMeta: makeMeta(cr, NodeDaemonSet),
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: defaultLabels(cr),
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: makeMeta(cr, NodeDaemonSet),
+				Spec: corev1.PodSpec{
+					PriorityClassName:  getPriorityClassName(cr),
+					ServiceAccountName: getNodeServiceAccountName(cr),
+					Containers: []corev1.Container{
+						{
+							Name:  "csi-node-driver-registrar",
+							Image: "quay.io/k8scsi/csi-node-driver-registrar:v1.2.0",
+							Args:  []string{"--v=5", "--csi-address=$(ADDRESS)", "--kubelet-registration-path=$(DRIVER_REG_SOCK_PATH)"},
+							Lifecycle: &corev1.Lifecycle{
+								PreStop: &corev1.Handler{
+									Exec: &corev1.ExecAction{Command: []string{"/bin/sh", "-c", "rm -rf /registration/linstor.csi.linbit.com /registration/linstor.csi.linbit.com-reg.sock"}},
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "ADDRESS",
+									Value: "/csi/csi.sock",
+								},
+								{
+									Name:  "DRIVER_REG_SOCK_PATH",
+									Value: "/var/lib/kubelet/plugins/linstor.csi.linbit.com/csi.sock",
+								},
+								{
+									Name: "KUBE_NODE_NAME",
+									ValueFrom: &corev1.EnvVarSource{
+										FieldRef: &corev1.ObjectFieldSelector{FieldPath: "spec.nodeName"},
```

Comment:
Too much nesting. Maybe move the `Container` construction into its own function?

Reply:
Split up the literal into smaller chunks with (hopefully) descriptive names. I kept them all in the same function as they are closely related anyways